### PR TITLE
Refractored workshop_app

### PR DESCRIPTION
**Pull request by pnshiralkar**
Major changes:
Upgraded Django to v.3.0.5
Upgraded Bootstrap to v4.4
Separate model for file attachments for workshop type (multiple files supported)

### CODE REVIEW

File: workshop_app/forms.py [hunk 1]
```diff
@@ -170,48 +91,47 @@ def save(self):
         new_profile = Profile(user=new_user)
         new_profile.institute = cleaned_data["institute"]
         new_profile.department = cleaned_data["department"]
-        #new_profile.position = cleaned_data["position"]
+        # new_profile.position = cleaned_data["position"]
```
Comment (ankitjavalkar) [inline workshop_app/forms.py:94]: Remove this line. It's not being used any way

File: workshop_app/forms.py [hunk 2]
```diff
@@ -223,48 +143,34 @@ def __init__(self, *args, **kwargs):
         self.fields['first_name'].initial = user.first_name
         self.fields['last_name'].initial = user.last_name
 
-class CreateWorkshop(forms.ModelForm):
+
+# debug : Changed this
```
Comment (ankitjavalkar) [inline workshop_app/forms.py:147]: Avoid leaving comments as much as is possible

File: workshop_app/forms.py [hunk 3]
```diff
@@ -170,48 +91,47 @@ def save(self):
...
         new_profile.title = cleaned_data["title"]
         new_profile.state = cleaned_data["state"]
-        new_profile.how_did_you_hear_about_us = cleaned_data["how_did_you_hear_about_us" ]
+        new_profile.how_did_you_hear_about_us = cleaned_data["how_did_you_hear_about_us"]
         new_profile.activation_key = generate_activation_key(new_user.username)
-        new_profile.key_expiry_time = timezone.now() + \
-                                    timezone.timedelta(days=1)
+        new_profile.key_expiry_time = timezone.now() + timezone.timedelta(days=1)
         new_profile.save()
         key = Profile.objects.get(user=new_user).activation_key
```
Comment (adityacp) [inline workshop_app/forms.py:103]: No need of doing this query again. Store the result obtained from `generate_activation_key` function and use the same wherever necessary

File: workshop_app/templates/workshop_app/activation.html [hunk 4]
```diff
@@ -1,23 +1,25 @@
...
-		</ul>
-	  	</div>
+    <nav class="navbar navbar-default navbar-custom">
+        <div class="container-fluid">
+            <div class="navbar-header">
+                <a class="navbar-brand" href="#">FOSSEE Python Workshops</a>
+            </div>
+            <ul class="nav navbar-nav navbar-right">
+                {% if request.user.profile.is_email_verified %}
+                    <li><a href="{{ URL_ROOT }}/view_profile/"><span class="glyphicon glyphicon-user"></span>
```
Comment (adityacp) [inline workshop_app/templates/workshop_app/activation.html:15]: Wherever `{{ URL_ROOT }}/view_profile/` this format is used, replace that with `{% url '<app_name>:<url_name>' args %}` format.
Make the above change wherever necessary.

File: workshop_app/templates/workshop_app/base.html [hunk 5]
```diff
@@ -1,100 +1,105 @@
...
-		{% endblock %}
-	</title>
-	<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
+    <meta name="viewport" content="width=device-width, initial-scale=1">
+    <title>
+        {% block title %}
+            HomePage
+        {% endblock %}
+    </title>
+    <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.4.1/css/bootstrap.min.css"
```
Comment (adityacp) [inline workshop_app/templates/workshop_app/base.html:10]: Download the bootstrap files locally and add their path in here. Adding CDN might lead to delays.

File: workshop_app/templates/workshop_app/base.html [hunk 6]
```diff
@@ -1,100 +1,105 @@
...
+          integrity="sha384-Vkoo8x4CGsO3+Hhxv8T/Q5PaXtkKtu6ug5TOeNV6gBiFeWPGFN9MuhOf23Q9Ifjh" crossorigin="anonymous">
 
-    <link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js">
-    <link href="https://ajax.googleapis.com/ajax/libs/jquery/3.1.1/jquery.min.js">
-    <link rel="stylesheet" href="{{ URL_ROOT }}/static/workshop_app/css/sticky-footer.css" type="text/css" />
+    <script src="https://code.jquery.com/jquery-3.4.1.slim.min.js"
```
Comment (adityacp) [inline workshop_app/templates/workshop_app/base.html:13]: Same here, download the JS, CSS files

File: workshop_app/templates/workshop_app/base.html [hunk 7]
```diff
@@ -1,100 +1,105 @@
...
+            integrity="sha384-Q6E9RHvbIyZFJoft+2mJbHaEWldlvI9IOYy5n3zV9zzTtmI3UksdQRVvoxMfooAo"
+            crossorigin="anonymous"></script>
+    {% block extra-dependencies %}
+    {% endblock %}
+    <script src="https://stackpath.bootstrapcdn.com/bootstrap/4.4.1/js/bootstrap.min.js"
+            integrity="sha384-wfSDF2E50Y2D1uUdj0O3uMBJnjuUD4Ih7YwaYd1iqfktj0Uod8GCExl3Og8ifwB6"
+            crossorigin="anonymous"></script>
+    <link href="https://fonts.googleapis.com/icon?family=Material+Icons"
+          rel="stylesheet">
+    <link rel="stylesheet" href="{{ URL_ROOT }}/static/workshop_app/css/base.css" type="text/css"/>
```
Comment (adityacp) [inline workshop_app/templates/workshop_app/base.html:26]: Same here, download the JS, CSS files

File: workshop_app/templates/workshop_app/base.html [hunk 8]
```diff
@@ -1,100 +1,105 @@
...
+    <script src="https://cdn.jsdelivr.net/npm/popper.js@1.16.0/dist/umd/popper.min.js"
+            integrity="sha384-Q6E9RHvbIyZFJoft+2mJbHaEWldlvI9IOYy5n3zV9zzTtmI3UksdQRVvoxMfooAo"
+            crossorigin="anonymous"></script>
+    {% block extra-dependencies %}
+    {% endblock %}
+    <script src="https://stackpath.bootstrapcdn.com/bootstrap/4.4.1/js/bootstrap.min.js"
+            integrity="sha384-wfSDF2E50Y2D1uUdj0O3uMBJnjuUD4Ih7YwaYd1iqfktj0Uod8GCExl3Og8ifwB6"
+            crossorigin="anonymous"></script>
+    <link href="https://fonts.googleapis.com/icon?family=Material+Icons"
+          rel="stylesheet">
```
Comment (adityacp) [inline workshop_app/templates/workshop_app/base.html:25]: Same here, download the JS, CSS files

File: workshop_app/templates/workshop_app/base.html [hunk 9]
```diff
@@ -1,100 +1,105 @@
...
+    <link rel="stylesheet" href="{{ URL_ROOT }}/static/workshop_app/css/base.css" type="text/css"/>
     <!-- favicon -->
-    <link rel="shortcut icon" type="image/png" href="{{ URL_ROOT}}/static/workshop_app/img/fevicon_python.png"/>
+    <link rel="shortcut icon" type="image/png" href="{{ URL_ROOT }}/static/workshop_app/img/fevicon_python.png"/>
```
Comment (adityacp) [inline workshop_app/templates/workshop_app/base.html:28]: For static files, use Django static URL format `{% static 'path-to-file' %}`.

File: workshop_app/templates/workshop_app/propose_workshop.html [hunk 10]
```diff
@@ -1,63 +1,71 @@
...
+                <div class="card-body">
+                    <div class="form-group">
+                        <form method="post">
+                            {% csrf_token %}
+                            <br/>
+                            {{ form.workshop_type }}
+                            <br>
+                            {{ form.date }}
+                            <br>
+                            {{ form.tnc_accepted }} I accept the <a href="/tnc">terms and conditions</a>
```
Comment (adityacp) [inline workshop_app/templates/workshop_app/propose_workshop.html:60]: Show the terms and conditions of the workshop type.

File: workshop_app/views.py [hunk 11]
```diff
@@ -1,112 +1,86 @@
...
+               "Prathamesh Salunke", "Kiran Kishore",
+               "KhushalSingh Rajput", "Prabhu Ramachandran",
+               "Arun KP"]
 
 
+# Helper functions
+
 def is_email_checked(user):
     if hasattr(user, 'profile'):
         return True if user.profile.is_email_verified else False
```
Comment (adityacp) [inline workshop_app/views.py:35]: Just return `user.profile.is_email_verified`

File: workshop_app/views.py [hunk 12]
```diff
@@ -1,112 +1,86 @@
...
     else:
         return False
 
 
-def is_superuser(user):
-    return True if user.is_superuser else False
+def is_instructor(user):
+    """Check if the user is having instructor rights"""
+    return True if user.groups.filter(name='instructor').count() > 0 else False
```
Comment (adityacp) [inline workshop_app/views.py:42]: Just return `user.groups.filter(name='instructor').exists()`

File: workshop_app/views.py [hunk 13]
```diff
@@ -1,112 +1,86 @@
...
 
 
-def index(request):
-    '''Landing Page'''
+def get_landing_page(user):
+    # For now, landing pages of both instructor and coordinator are same
+    if user.groups.filter(name='instructor').count() > 0:
```
Comment (adityacp) [inline workshop_app/views.py:47]: use `is_instructor` function here.

File: workshop_app/views.py [hunk 14]
```diff
@@ -115,1178 +89,290 @@ def user_login(request):
...
 
 def activate_user(request, key=None):
     user = request.user
-    if is_superuser(user):
+    if user.is_superuser:
         return redirect("/admin")
     if key is None:
-        if user.is_authenticated() and user.profile.is_email_verified==0 and \
-        timezone.now() > user.profile.key_expiry_time:
+        if user.is_authenticated and user.profile.is_email_verified == 0 and \
```
Comment (adityacp) [inline workshop_app/views.py:102]: No need to do equal to 0.

File: workshop_app/views.py [hunk 15]
```diff
@@ -115,1178 +89,290 @@ def user_login(request):
...
+                timezone.now() > user.profile.key_expiry_time:
             status = "1"
             Profile.objects.get(user_id=user.profile.user_id).delete()
             User.objects.get(id=user.profile.user_id).delete()
             return render(request, 'workshop_app/activation.html',
-                        {'status':status})
-        elif user.is_authenticated() and user.profile.is_email_verified==0:
+                          {'status': status})
+        elif user.is_authenticated and user.profile.is_email_verified == 0:
```
Comment (adityacp) [inline workshop_app/views.py:109]: Same here

File: workshop_app/views.py [hunk 16]
```diff
@@ -115,1178 +89,290 @@ def user_login(request):
...
-    except:
-        return redirect('/register/')
-
-    if key == user.activation_key:
-        user.is_email_verified = True
-        user.save()
-        status = "0"
+    user = Profile.objects.filter(activation_key=key)
+    if user.count() > 0:
+        user = user[0]
```
Comment (adityacp) [inline workshop_app/views.py:120]: Use `user.exists()` here and then instead of user[0] use `user.first()`

File: workshop_app/views.py [hunk 17]
```diff
@@ -115,1178 +89,290 @@ def user_login(request):
...
-                        )
-
-                    #For Coordinator
-                    send_email(request, call_on='Change Date',  
-                        new_workshop_date=str(new_workshop_date.date()),
-                        workshop_date=str(workshop_date.date()),
-                        other_email=cemail.email
-                        )
+                        workshop = Workshop.objects.filter(id=client_data.get('workshop_id'))
+                        workshop_date = workshop[0].date
```
Comment (adityacp) [inline workshop_app/views.py:255]: Instead of `workshop[0].date` use `workshop.first().date`

File: workshop_app/views.py [hunk 18]
```diff
@@ -115,1178 +89,290 @@ def user_login(request):
...
-
-                    #For Coordinator
-                    send_email(request, call_on='Booking Request Rejected',
-                        workshop_date=str(client_data[1]),
-                        workshop_title=wtitle,
-                        other_email=cmail
-                        )
-
-            workshops = []
             today = datetime.today().date()
```
Comment (adityacp) [inline workshop_app/views.py:272]: You can use timezone.now() from Django utils.

File: workshop_app/views.py [hunk 19]
```diff
@@ -115,1178 +89,290 @@ def user_login(request):
...
+                    return redirect('/my_workshops/')
+                else:
+                    form_data.save()
+                    instructors = Profile.objects.filter(position='instructor')
+                    for i in instructors:
+                        send_email(request, call_on='Proposed Workshop',
+                                   user_position='instructor',
+                                   workshop_date=str(form_data.date),
+                                   workshop_title=form_data.workshop_type,
+                                   user_name=str(user.get_full_name()),
```
Comment (adityacp) [inline workshop_app/views.py:331]: No need of doing `str`, `user.get_full_name()` is string by default.

File: workshop_app/views.py [hunk 20]
```diff
@@ -115,1178 +89,290 @@ def user_login(request):
...
-                                )
-
-            upcoming_workshops = []
-            for workshop in proposed_workshops:
-                upcoming_workshops.append(workshop)
-
-            for workshop in requested_workshops:
-                upcoming_workshops.append(workshop)
+        # If page is out of range(e.g 999999), deliver last page.
+        workshop_type = paginator.page(paginator.num_pages)
```
Comment (adityacp) [inline workshop_app/views.py:376]: You can use `paginator.get_page`  instead of doing try-except every time.

File: workshop_portal/urls.py [hunk 21]
```diff
@@ -27,7 +27,7 @@
 urlpatterns = [
     url(r'^admin/', admin.site.urls),
     url(r'^', include('workshop_app.urls')),
-    url(r'^', include('workshop_app.urls_password_reset')),
-    url(r'^', include('statistics_app.urls')),
+    # url(r'^', include('workshop_app.urls_password_reset')),
+    # url(r'^', include('statistics_app.urls')),
```
Comment (adityacp) [inline workshop_portal/urls.py:31]: Do not add commented code.

File: workshop_app/templates/workshop_app/base.html (hunk 9 above)
Comment (ankitjavalkar) [reply workshop_app/templates/workshop_app/base.html:28]: Will be fixing in a later PR

File: workshop_app/templates/workshop_app/base.html (hunk 8 above)
Comment (ankitjavalkar) [reply workshop_app/templates/workshop_app/base.html:25]: Will be fixing in a later PR

File: workshop_app/templates/workshop_app/base.html (hunk 5 above)
Comment (ankitjavalkar) [reply workshop_app/templates/workshop_app/base.html:10]: Will be fixing in a later PR

File: workshop_app/templates/workshop_app/activation.html (hunk 4 above)
Comment (ankitjavalkar) [reply workshop_app/templates/workshop_app/activation.html:15]: Will be fixing in a later PR

**Comment by ankitjavalkar** [comment]
- [ ] Fix the terms and conditions link
- Display form errors when booking workshops - if date field is blank and/or if T&C boolean field is blank
- [ ] Coordinator profile template has been deleted, replace or fix it, so that Instructor can view the coordinator profile
- [ ] Workshop Name is not visible in the list after workshop is proposed

File: workshop_app/templates/workshop_app/propose_workshop.html (hunk 10 above)
Comment (pnshiralkar) [reply workshop_app/templates/workshop_app/propose_workshop.html:62]: Will be fixing later

**Comment by ankitjavalkar** [comment]
Merging now.

**Comment by CruiseDevice** [comment]
Closes #111

**Pull request merged**